include/open62541_compat.h

```diff
@@ -47,7 +47,11 @@
 #include <opcua_basedatavariabletype.h>
 
 
-
+#define OPEN62541_COMPAT_LOG_AND_THROW(EXCEPTION_TYPE, MSG) \
+    { \
+    LOG(Log::ERR) << MSG; \
```

Comment:
LogIt's LOG macro exposes a std::ostream, allowing client code to << put << stuff << in << like that (by instantiating a LogRecord object and exposing its internal std::ostream member).

Could something like that work here too? So client code "sees" a std::ostream and uses the usual stream operators to pump info into the eventual LOG call and whatever exception type is eventually thrown.

And if not, is there any benefit in using a macro - I guess template function would serve equally well

Reply:
> Just an observation: no capacity for LogIt components

yes, it's not there, and technically this feature does not prevent it.
Why it's not there (and perhaps you should comment on it) is that adding open62541-wide components (a good idea, I think) would require that open62541-compat users initialize LogIt beforehand. It seems not so elegant to me to force it and the consequences of using LogIt components without expecting someone initializes them are quite grave to my mind. What's your idea on it?

Reply:
> LogIt's LOG macro exposes a std::ostream, allowing client code to << put << stuff << in << like that (by instantiating a LogRecord object and exposing its internal std::ostream member).
> 
> Could something like that work here too? So client code "sees" a std::ostream and uses the usual stream operators to pump info into the eventual LOG call and whatever exception type is eventually thrown.
> 
> And if not, is there any benefit in using a macro - I guess template function would serve equally well

Would you be interested in accepting the offer of an improvement ticket for it?

Reply:
Not sure if I follow, but maybe some facts might help...

You can call register for a logging component at any time. If the LogIt instance (the central object that stores the verbosity mapping for each component) does not exist then register fails - returns Log::INVALID_HANDLE. That LogIt instance object is created on LogIt initialisation.

If you (successfully) register the same component multiple times, the client code always receives the same handle. So, there's no danger (apart from efficiency) in registering some open62541-compat component, even if it was already registered elsewhere.

If using LogIt in a quasar server - it's nice to register IDs before Meta runs, so that components are represented in the AS and can be controlled via config/runtime.

Something useful there maybe?

Reply:
Yes, within a quasar server that (probably) wouldn't be a problem and obviously could be done within quasar.
But what do we do with (now massive number of) people using open62541-compat out of quasar servers?
They would have to add to their applications, at init time, the statement to register this component; not doing so would disable them from *ANY* LogIt output originating from open62541-compat, correct?

Reply:
Nope, don't think so anyway... 
* You have to initialise LogIt in order to successfully register a component
* You have to register a component in order to successfully use it in a LOG statement. 

But that's all constraint-wise.

Reply:
So you say that logging to a component ID that is not registered will result in a log message? What is the *implied* log level then?

Reply:
Nope: Call LOG with non-registered component --> no log message

Reply:
exactly what I thought. So in my opinion, to be respectful to different users out there, we can't immediately introduce separate component ID for open62541-compat, though I think it would be very useful overall. If you don't agree please reply, otherwise I'd consider this comment resolved.